[Frontend] [Core] feat: Add model loading using `tensorizer`

sangstar opened the pull request:
[Feature Request Issue
](https://github.com/vllm-project/vllm/issues/3533)
# Tensorizer Support
This PR allows models used for the OpenAI-compatible API server to be loaded using
[Coreweave's Tensorizer](https://github.com/coreweave/tensorizer), enabling extremely fast (faster than cached Safetensors) model loads from HTTP/HTTPS, 
Redis, and S3 endpoints.

The key changes involved are:

1. Adds `TensorizerConfig` to the set of configs.
2. Adds `tensorizer_loader.py` to `vllm/model_executor` that provides utility functions
for tensorizer. 
3. Adds multiple args to the vLLM's OpenAI inference service entrypoint that allows
the user to specify the path to serialized-by-tensorizer model tensors, as well as 
arguments for tensorizer's deserializer. 
4. Allows deserialization of serialized model tensors in HuggingFace's model format, 
as well as supporting deserializing serialized vLLM-formatted models, allowing the 
use of loading with `plaid_mode`, which can allow Llama 2 13B non-locally to start serving requests
in as little as 10 seconds. Also supports encrypting and decrypting model tensors.
5. Adds a `tensorize_vllm_model.py` script to `examples/` that allows vLLM models to be serialized and
deserialized with `tensorizer`.
6. Adds `tensorizer` as an optional dependency.

Credentialing for S3 is supported by passing a user's access and secret key to 
`S3_ACCESS_KEY_ID` and `S3_SECRET_ACCESS_KEY` environment variables respectively. It can also be specified as CLI args to the api server entrypoint.

# Model loading benchmarks

Tensorizer can load models like Llama 2 13B in as little as 10 seconds. In order to do so, a model must be 
serialized using `TensorSerializer` to a `.tensors` file located either locally or through a S3, HTTP/HTTPS, or Redis
endpoint. `--tensorizer-uri` must be specified with the serialized tensors location when invoking the API server. 

Example usage:

```
python -m vllm.entrypoints.openai.api_server \
--host 0.0.0.0 \
--model EleutherAI/pythia-6.9b \
--load-format tensorizer \
--tensorizer-uri s3://tensorized/EleutherAI/pythia-6.9b/fp16/model.tensors

```

If a vLLM model is serialized, `plaid_mode` can be used, which loads much faster. The following plot demonstrates model loading time benchmarks for vLLM's OpenAI-compatible inference server on a Nvidia A40 GPU.

Tensorizer is so fast that it loads models faster than Safetensors _even locally._

![Benchmark (10)](https://github.com/vllm-project/vllm/assets/51196172/89b02c4b-05c9-4f43-83a3-22802a7a91d0)

sangstar committed:
feat: Support loading model tensors using `tensorizer`

This feature allows vLLM models to be loaded extremely fast using
`tensorizer`. `tensorizer` loads serialized model tensors from
HTTP/HTTPS, Redis, S3 endpoints, or locally, typically on the scale
of multiple GB/s.

sangstar committed:
fix: Remove unnecessary files

sangstar committed:
fix(vllm-tensorizer): Allow providing S3 credentials

This allows the deserializer to access S3 credentials when reading.
It extracts the access key from env variables `S3_ACCESS_KEY_ID`
and secret key from `S3_SECRET_ACCESS_KEY`.

sangstar committed:
fix: Fix passing S3 auth vars through stream

Previous commit wasn't able to pass S3
credentials through `TensorDeserializer`, so
it is instead passed to `stream_io.openstream`
which is used to instantiate the
`TensorDeserializer`.

sangstar committed:
fix: Disallowing `plaid_mode = False` and updating `tensorizer` version

Removed functionality to allow for tensorizing without `plaid_mode`,
and updating to `tensorizer==2.8.0`

sangstar committed:
refactor: Retire use of `download_dir` as `TensorizerArgs` param

Replaces `download_dir` with `tensorizer_uri` as a `TensorizerArgs`
param. This is due to discussions on `download_dir` being a confusing
and ultimately not helpful parameter for the location of model tensors.
Instead, `download_dir` is back to the definition coinciding with
the convention set by HuggingFace; as a location to download weights
for caching. A new parameter takes its place: `tensorizer_uri`. This
specifically deals with locating model tensors for `tensorizer`.

sangstar committed:
fix: Remove `store_true` action for `--tensorizer-uri`

Also slight formatting fix for warning when loading weights with
`download_dir` not set to `None`.

sangstar committed:
refactor: No 2x copying for `tensorizer` (WIP)

sangstar committed:
chore: Omit commandeering weight loaders for merging layers (WIP)

sangstar committed:
feat: Re-add deserializing vLLM models

Integrated changes from
ssteel/tensorizer-support branch that allowed
for deserializing vLLM models.

sangstar committed:
chore: Harmonize CPU and GPU deserializing

Integrated previous support for vLLM-formatted
model loading with `tensorizer` that makes full
use of loading to the GPU with `plaid-mode`, as
well as falling back on being able to load
HuggingFace models for serving using the CPU so
that vLLM can perform its manual GPU loading.

sangstar committed:
perf: Add `force_http=True` for faster loading speeds

sangstar committed:
chore: Reformat code with `format.sh`, cleanup debugging code

sangstar committed:
chore: Fix formatting, some misc. changes

Fixed some unnecessary formatting changes in `arg_utils.py`,
`weight_utils.py` and `model_loader.py` fixed improperly passing
`force_http` to `TensorDeserializer` rather than `open_stream`.

sangstar committed:
fix: Correct logging for loading tensorizer with cpu

sangstar committed:
chore: Implement changes from feedback

Misc. fixes from now resolved conversations. Mostly consisting of
changes to syntax, style, adding docstrings, and versioning.

sangstar committed:
fix: Correctly instantiate vLLM-formatted models

`examples/tensorize_vllm_model.py` now correctly instantiates
vLLM-formatted models.

sangstar committed:
chore: Reformat and delete deprecated comment from `.ipynb`

sangstar committed:
perf: Allow passing of deserializer args from `TensorizerArgs`

sangstar committed:
style: Reformat with new formatting changes

sangstar committed:
Run yapf and ruff

Refined formatting to adhere to vLLM's specifications in `format.sh`
Commit message neglecting conventional commit templating due to
vLLM's preference for commit messaging on using their formatter.

sangstar committed:
fix: Fix incorrect `TensorizerArgs` import in `config.py`

Also enforces a default float16 precision for models when serializing

sangstar committed:
perf: Multiple misc. improvements from code review

Multiple fixes and performance improvements from code reviews.
Some fixes include adjusting the `tensorizer` version, an overhaul
of tensorize_vllm_model.py to allow for encryption as well as other
improvements, and some slight fixes to tensorizer_loader.py

sangstar committed:
pref: More misc. fixes to complete initial code review

A multitude of fixes requested from code reviewers have been
implemented, ranging from encryption support, more logical handling
of tensorizer parameters and argparsing, documentation syntax fixes,
and some function fixes

sangstar committed:
fix: Remove `print(tensorizer_args)`

sangstar committed:
Run yapf and ruff

sangstar committed:
fix: Add specific category for warnings with `PerformanceWarning`

sangstar committed:
chore: Multiple fixes from final code review

Multiple fixes have been made mostly pertaining to arg parsing and
docs fixes from final code review. Some changes include the changing
of `tensorizer_args` to a dynamic attribute for `EngineArgs`, as well
as multiple QoL changes for example script.

sangstar committed:
fix: Add `s3_endpoint` as attr for `TensorizerArgs`

sangstar committed:
chore: Remove `filter_func` from CLI args, some doc fixes

sangstar committed:
chore: Allow env var or CLI arg specification for S3 credentials

sangstar committed:
fix: Disallow using `force_http`

sangstar committed:
chore: Remove unnecessary print statement in example script

sangstar committed:
Run yapf and ruff

sangstar committed:
Run yapf and ruff

Also added test files

sangstar committed:
Run yapf and ruff

sangstar committed:
docs: Update `tensorizer` as a `--load-format` in `engine_args.rst`

sangstar committed:
fix: Restore `tensorizer_args` as instance attr to `EngineArgs`

Brings back `tensorizer_args` as an instance attr to `EngineArgs`
and refactors `EngineArgs.from_cli_args` to correctly parse and
instantiate `EngineArgs` and `TensorizerArgs` while preserving
`tensorizer_args` as an attr.

sangstar committed:
Run yapf and ruff

sangstar committed:
chore: Move testing out of own test folder

sangstar committed:
fix: Add `tensorizer >= 2.8.1` to `requirements-rocm.txt` for CI

sangstar committed:
fix: Add version of `tensorizer` that will pass testing suite

Adds a branched version of `tensorizer` that doesn't init CUDA
so as to not init CUDA before forking during tests

sangstar committed:
chore: Add notice that `requirements-dev` dep can be removed `>2.8.1`

Adds comment that the specific `tensorizer` dep for `requirements-dev`
can be dropped after `tensorizer > 2.8.1` releases

sangstar committed:
fix: Resolve double `HfFileSystem` import

sangstar committed:
style: Run `isort`

sangstar commented:
@cadedaniel @rkooo567 Pinging for an assigned reviewer from someone on the team when possible!

sangstar committed:
Run yapf and ruff

sangstar committed:
fix: Add `tensorizer` to mock imports

sangstar committed:
perf: Add newest `tensorizer` version that will not init CUDA

sangstar committed:
fix: Adjust `tensorizer` version for `requirements-dev.txt`

sangstar committed:
chore: Rebase and fix carrying over changes to `arg_utils` typing

sangstar committed:
fix: Add `tensorizer` to `requirements-cpu.txt`

sangstar commented:
@cadedaniel @rkooo567 @Yard1 @WoosukKwon @zhuohan123 

All tests are passing. Can I get eyes on this please? Cheers!

sangstar committed:
perf: Add concurrent reading to `TensorDeserializer`

sangstar committed:
docs: Add `num_readers` docstring

sangstar committed:
chore: Replace `PerformanceWarning` after rebase

sangstar committed:
Run yapf and ruff

sangstar commented:
Hello! Just pinging the review team for an assigned reviewer. This would be a hugely impactful feature for vLLM users, especially those who prioritize responsiveness during inference, tensor encryption during loading, or extremely fast model loading via HTTP! Please refer to my benchmarks here and the [feature request I made](https://github.com/vllm-project/vllm/issues/3533) for more on that.

@cadedaniel @rkooo567  @Yard1 @WoosukKwon @zhuohan123  @ywang96 
@esmeetu @hmellor @liangfu  @LiuXiaoxuanPKU  @merrymercy @njhill @pcmoritz @simon-mo @suquark @Ying1123 @youkaichao @ywang96 @zhisbug

ywang96 commented:
Hey @sangstar ! Thank you for the contribution - I will take a look at this PR next week and test it out!

Yard1 commented on requirements.txt:
```diff
@@ -7,6 +7,7 @@ numpy
 torch == 2.1.2
 requests
 py-cpuinfo
+tensorizer == 2.9.0a0
```
this shouldn't become a mandatory dependency

Yard1 commented on vllm/model_executor/weight_utils.py:
```diff
@@ -277,6 +287,24 @@ def hf_model_weights_iterator(
             with open(param_path, "rb") as f:
                 param = np.load(f)
             yield name, torch.from_numpy(param)
+    elif load_format == "tensorizer":
+        tensorizer_args = load_format.params
+        warnings.warn(
+            "Deserializing HuggingFace models is not optimized for "
+            "loading on vLLM, as tensorizer is forced to load to CPU. "
+            "Consider deserializing a vLLM model instead for faster "
+            "load times. See the examples/tensorize_vllm_model.py example "
+            "script for serializing vLLM models.",
+            category=PerformanceWarning,
+            stacklevel=2)
+        deserializer_args = tensorizer_args.deserializer_params
+        stream_params = tensorizer_args.stream_params
+        stream = open_stream(tensorizer_args.tensorizer_uri, **stream_params)
+        with TensorDeserializer(stream, **deserializer_args,
+                                device="cpu") as state:
+            for name, param in state.items():
+                yield name, param
+        del state
```
This can be done in a followup, but it would be good to refactor the weight_utils to use some sort of an interface for different weight loading methods instead of just having multiple branches

sangstar replied:
I agree, and will assume this is out of the scope of this PR.

Yard1 commented on vllm/engine/arg_utils.py:
```diff
@@ -409,9 +415,16 @@ def add_cli_args(
     def from_cli_args(cls, args: argparse.Namespace) -> 'EngineArgs':
         # Get the list of attributes of this dataclass.
-        attrs = [attr.name for attr in dataclasses.fields(cls)]
+        attrs = [
+            attr.name for attr in dataclasses.fields(cls)
+            if attr.name != "tensorizer_args"
+        ]
         # Set the attributes from the parsed arguments.
         engine_args = cls(**{attr: getattr(args, attr) for attr in attrs})
+
+        engine_args.tensorizer_args = TensorizerArgs.from_cli_args(args) if (
+            engine_args.load_format == "tensorizer") else None
+
```
this is pretty hacky, maybe it would be a better idea to move this logic to `create_engine_config`

sangstar replied:
This is a bit tricky as I implemented this in a way where `tensorizer_args` is expected to be an `attr` of `EngineArgs` but not an arg for its argparser, which didn't feel great but seemed to be a necessary evil outside of creating an entirely new `TensorizerConfig`. I could propose a different solution that removes `tensorizer_args` from `EngineArgs` but is also slightly hacky.

In the `__post_init__` for `EngineArgs`..
```
    def __post_init__(self):
        if self.tokenizer is None:
            self.tokenizer = self.model
+      if not hasattr(self, 'tensorizer_args'):
+          self.tensorizer_args = None
```

With a perhaps slightly nicer `EngineArgs.from_cli_args`:

```
    @classmethod
    def from_cli_args(cls, args: argparse.Namespace) -> 'EngineArgs':
        # Get the list of attributes of this dataclass.
        attrs = [attr.name for attr in dataclasses.fields(cls)]
        # Set the attributes from the parsed arguments.
        engine_args = cls(**{attr: getattr(args, attr) for attr in attrs})

        # Check if tensorizer_args are expected.
        if args.load_format == "tensorizer":
            tensorizer_attrs = [
                attr.name for attr in dataclasses.fields(TensorizerArgs)
            ]
            # Create an instance of TensorizerArgs with its subset of the CLI args.
            engine_args.tensorizer_args = TensorizerArgs(
                **{attr: getattr(args, attr) for attr in tensorizer_attrs}
            )

        return engine_args
```

And a hacky tweak to `LLM`'s `__init__`:

```
        if "disable_log_stats" not in kwargs:
            kwargs["disable_log_stats"] = True
+      if "tensorizer_args" in kwargs:
+          tensorizer_args = kwargs["tensorizer_args"]
+          del kwargs["tensorizer_args"]
+      else:
+          tensorizer_args = None
        engine_args = EngineArgs(
            model=model,
            tokenizer=tokenizer,
            tokenizer_mode=tokenizer_mode,
            trust_remote_code=trust_remote_code,
            tensor_parallel_size=tensor_parallel_size,
            dtype=dtype,
            quantization=quantization,
            revision=revision,
            tokenizer_revision=tokenizer_revision,
            seed=seed,
            gpu_memory_utilization=gpu_memory_utilization,
            swap_space=swap_space,
            enforce_eager=enforce_eager,
            max_context_len_to_capture=max_context_len_to_capture,
            disable_custom_all_reduce=disable_custom_all_reduce,
            **kwargs,
        )
+      engine_args.tensorizer_args = tensorizer_args
        self.llm_engine = LLMEngine.from_engine_args(
            engine_args, usage_context=UsageContext.LLM_CLASS)
        self.request_counter = Counter()
```

Yard1 replied:
Let me think about this!

Yard1 commented:
Out of curiosity, how does support for tensor parallelism sharded models is looking like with tensorizer?

sangstar commented on examples/tensorize_vllm_model.py:
```diff
@@ -0,0 +1,230 @@
+    initialize_model_parallel)
+from vllm.model_executor.tensorizer_loader import TensorizerArgs
+
+
```
Yes! Can do.

sangstar commented on vllm/model_executor/model_loader.py:
```diff
@@ -77,7 +80,11 @@ def get_model(model_config: ModelConfig, device_config: DeviceConfig,
         # The weights will be initialized as empty tensors.
         with torch.device(device_config.device):
-            if hasattr(model_class, "supported_lora_modules"):
+            if model_config.load_format == "tensorizer" and _is_vllm_model(
+                    model_config):
+                model = load_with_tensorizer(model_class, model_config)
+                return model.eval()
+            elif hasattr(model_class, "supported_lora_modules"):
                 model = model_class(model_config.hf_config, linear_method,
                                     lora_config)
```
Using Tensorizer and LoRA should be able to both be used.

The code block you're highlighting is a specific instance of using Tensorizer using `load_with_tensorizer`, which skips loading a model through its instantiation as a vLLM model (not loading a model from HuggingFace and loading it with vLLM's bespoke layer modules). It assumes that the model to be deserialized _is itself already_ a vLLM model. My example script `examples/tensorize_vllm_model.py` shows an example of serializing a model at a state that is ready to do inference on vLLM. I'm not sure how that interacts with LoRA adapters in that case.

Tensorizer can additionally be invoked the "standard" way, like any other load format, through its implementation in `weight_utils.py`, and using serialized models from the HF model hub. In this instance, LoRA adapters should still work, as it's simply loading in the tensors and handing it off to vLLM's `weight_loader`s as normal. Looking at your example script for multiLoRA inference, I was able to successfully run it using Tensorizer with this modification:

```
def initialize_engine() -> LLMEngine:
    """Initialize the LLMEngine."""
    # max_loras: controls the number of LoRAs that can be used in the same
    #   batch. Larger numbers will cause higher memory usage, as each LoRA
    #   slot requires its own preallocated tensor.
    # max_lora_rank: controls the maximum supported rank of all LoRAs. Larger
    #   numbers will cause higher memory usage. If you know that all LoRAs will
    #   use the same rank, it is recommended to set this as low as possible.
    # max_cpu_loras: controls the size of the CPU LoRA cache.
    from vllm.model_executor.tensorizer_loader import TensorizerArgs
    engine_args = EngineArgs(model="meta-llama/Llama-2-7b-hf",
        load_format="tensorizer",
        enable_lora=True,
        tensorizer_args=TensorizerArgs(
            tensorizer_uri=tensorizer_uri,
            num_readers=1),
        max_loras=1,
        max_lora_rank=8,
        max_cpu_loras=2,
        max_num_seqs=50 ## Lowered for GPU usage,
        max_model_len=1000 ## Lowered for GPU usage)
```

Yard1 replied:
Thanks! Just to be clear, does it mean that in an ideal case, we'll load directly to GPU?

For LoRA, the underlying layers have exactly the same weights, the only difference is that the LoRA layers instantiate some extra tensors (there is no extra data being loaded).

sangstar replied:
> Just to be clear, does it mean that in an ideal case, we'll load directly to GPU?

Yes, ideally Tensorizer can be used to have CUDA be the output device. The loading mechanism for vLLM seems to be initializing the model entirely in CPU memory, and then copying it into GPU memory. `tensorizer` achieves this much faster and with far less CPU memory usage by instantiating a few reusable CPU buffers, and the model with meta tensors when copying to CUDA devices rather than instantiating everything first into memory.

Yard1 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,249 @@
+    return tensorizer.deserialize()
+
+
+def _is_vllm_model(model_config: ModelConfig = None,
+                   file_uri: Optional[str] = None) -> bool:
+    if file_uri:
+        return "vllm" in file_uri
+    else:
+        return "vllm" in model_config.tensorizer_args.tensorizer_uri
```
this is pretty hacky and magical, I think it would be better to instead have users determine this through an argument (in `tensorizer_args`)

Yard1 commented:
Also, I tested this using the following workflow:
1. Use the provided script to generate local files for `meta-llama/Llama-2-7b-chat-hf`
2. Modify `examples/offline_inference.py` with following:
```python
llm = LLM(model="meta-llama/Llama-2-7b-chat-hf",
          tensor_parallel_size=1,
          load_format="tensorizer",
          tensorizer_args=TensorizerArgs(
              tensorizer_uri="/PATH/TO/vllm/model.tensors",
              num_readers=1,
          ),
          trust_remote_code=True)
```
And I am getting gibberish output.

sangstar commented:
> Out of curiosity, how does support for tensor parallelism sharded models is looking like with tensorizer?

Tensorizer currently doesn't support tensor parallelism sharded models, but tensors can currently be loaded into one GPU and then transferred device-to-device if there is space. 


> Also, I tested this using the following workflow:
> 
> 1. Use the provided script to generate local files for `meta-llama/Llama-2-7b-chat-hf`
> 2. Modify `examples/offline_inference.py` with following:
> 
> ```python
> llm = LLM(model="meta-llama/Llama-2-7b-chat-hf",
>           tensor_parallel_size=1,
>           load_format="tensorizer",
>           tensorizer_args=TensorizerArgs(
>               tensorizer_uri="/PATH/TO/vllm/model.tensors",
>               num_readers=1,
>           ),
>           trust_remote_code=True)
> ```
> 
> And I am getting gibberish output. Any idea what's wrong?
> 
> We should definitely test this code path, and ensure weights are identical between safetensors and tensorizer.

On my side my output doesn't seem completely nonsensical, and this is with a altered-to-be-lower max context length:

```
Prompt: 'Hello, my name is', Generated text: ' Dustin Nelson and I’m going to be your tutor!\n'
Prompt: 'The president of the United States is', Generated text: ' a member of an exclusive club.\nDonald Trump is not the only person to'
Prompt: 'The capital of France is', Generated text: ' one of the most visited cities in Europe, and its location makes it a great'
Prompt: 'The future of AI is', Generated text: ' all about interoperability\nHere’s why developers and businesses will start'
```

Perhaps it has to do with your serialized model tensors? Make sure to distinguish between the working definitions of a vLLM-formatted model and a HF-formatted model here. If you're trying to deserialize a vLLM model, make sure that its `tensorizer_uri` has `vllm` in it somewhere. Otherwise, if you're trying to deserialize a HF model, it shouldn't have `vllm` in the `tensorizer_uri` (I understand that this is a bit hacky, as per your comment, so I'll change that). The model also needs to be serialized beforehand. My example script shows how to serialize a vLLM-formatted model. Serializing a model from the HF model hub is also straight-forward, as pulled from the `README.md` for [Tensorizer](https://github.com/coreweave/tensorizer):

```
import torch
from tensorizer import TensorSerializer
from transformers import AutoModelForCausalLM

model_ref = "EleutherAI/gpt-j-6B"
# For less intensive requirements, swap above with the line below:
# model_ref = "EleutherAI/gpt-neo-125M"
model_name = model_ref.split("/")[-1]
# Change this to your S3 bucket.
s3_bucket = "bucket"
s3_uri = f"s3://{s3_bucket}/{model_name}.tensors"

model = AutoModelForCausalLM.from_pretrained(
    model_ref,
    revision="float16",
    torch_dtype=torch.float16,
    low_cpu_mem_usage=True,
)

serializer = TensorSerializer(s3_uri)
serializer.write_module(model)
serializer.close()
```

I'm happy to write a test confirming that the tensors aren't altered, although I'm not sure how painful that will be for the testing suite.

Yard1 commented:
@sangstar I ran
`python tensorize_vllm_model.py --model meta-llama/Llama-2-7b-chat-hf serialize --serialized-directory /PATH/TO/vllm` and then used the created on-disk file to load the model with 
```python
llm = LLM(model="meta-llama/Llama-2-7b-chat-hf",
          tensor_parallel_size=1,
          load_format="tensorizer",
          tensorizer_args=TensorizerArgs(
              tensorizer_uri="/PATH/TO/vllm/model.tensors",
              num_readers=1,
          ),
          trust_remote_code=True)
```

This is what I am getting as output:
```
Prompt: 'Hello, my name is', Generated text: ''
Prompt: 'The president of the United States is', Generated text: 'ajeguskyExcelристи upload Richmondegu Microsoft Hier Excel Nilimen arr intent'
Prompt: 'The capital of France is', Generated text: ' Offiznofollow Krit KritConfigurationonnenigeristiquedefin awkbibliothek Convertweenстваording'
Prompt: 'The future of AI is', Generated text: ''
```

For comparison, this is what I get without tensorizer:
```
Prompt: 'Hello, my name is', Generated text: " Sherry and I'm a 35-year-old woman from"
Prompt: 'The president of the United States is', Generated text: ' a member of Congress, which means that he or she is subject to the same'
Prompt: 'The capital of France is', Generated text: ' Paris. This is a fact that is well known and widely accepted. However,'
Prompt: 'The future of AI is', Generated text: ' likely to be shaped by a combination of technological advancements, soci'
```

also digging a little deeper into it and I am not sure about RotaryEmbedding anymore. I thought that was it as it was generated when model class was initialized, but it looks like whatever is loaded should be identical. I compared the weight tensors after loading with tensorizer and without and the model looks to be identical, but the outputs are different.

Yard1 commented:
I am getting gibberish with `EleutherAI/gpt-j-6B` as well. Let me know if there is any way I can help.

I am running unmodified code from this PR in a fresh environment on a g5.12xlarge instance.

Yard1 commented:
I figured it out, it's an issue with `linear_weights` attribute present on all of vLLM's linear layers.

Let's add e2e tests with a small model (like opt 125m) that ensure the output is exactly the same with and without tensorize.

sangstar commented:
> I figured it out, it's an issue with `linear_weights` attribute present on all of vLLM's linear layers. I will make a separate PR to fix it. The workaround for this PR is to add
> 
> ```python
>         for child in self.model.modules():
>             if hasattr(child, "linear_weights"):
>                 for name, weight in child.linear_weights.items():
>                     if isinstance(weight, torch.Tensor):
>                         child.linear_weights[name] = getattr(child, name)
> ```
> 
> in `TensorizerAgent.deserialize()` after deserializing the weights.
> 
> Let's add e2e tests with a small model (like opt 125m) that ensure the output is exactly the same with and without tensorizer.

Was currently working on this too. Thanks for that great catch! I'll add a test for this as you mentioned.

rkooo567 commented on examples/tensorize_vllm_model.py:
```diff
@@ -0,0 +1,230 @@
+from vllm.model_executor.tensorizer_loader import TensorizerArgs
+
+
+def parse_args():
```
So in order to use the tensorizer, we should serialize the model using this script and store it in S3 or local file?  Is it possible to add e2e example somewhere? (like I can follow line by line)? I think it is important to make serializing the model part straightforward and easy.

sangstar replied:
Yep Let me know if this recent docstring addition answers your question :)

https://github.com/vllm-project/vllm/pull/3476/commits/71a9f795b865b296c9ce34b1e3b669fa8bc36580#diff-b1567039ae60823bd62b460014416422145b1cf46be70d7265c66af6cfcaf974R22-R55

rkooo567 replied:
How do you usually set the credentials for S3? Is it a part of the API? Can you also add a short docstring to explain it?

sangstar committed:
fix: Fix model output on deserialization and add e2e output test

sangstar committed:
fix: Properly ensure test outputs are deterministic, add HF model test

sangstar commented on tests/test_tensorizer.py:
```diff
@@ -0,0 +1,54 @@
+
+
+@patch('vllm.model_executor.tensorizer_loader.TensorizerAgent')
+def test_load_with_tensorizer(mock_agent, model_config):
```
I have additional tests that now confirm that the outputs of models before and after serializing + deserializing are identical using greedy sampling. Do you want me to make an additional test for the weight loading or are you happy with that?

sangstar commented on tests/test_tensorizer.py:
```diff
@@ -0,0 +1,54 @@
+    assert result is True
+
+
+def test_is_vllm_model_without_vllm_in_uri(model_config):
```
I have two tests that verify that model outputs are identical before and after serializing then loading with tensorizer in `test_tensorizer.py`. Do you feel comfortable with those tests accounting for that?

sangstar committed:
fix: Make vLLM tensorizing specification less hacky

sangstar committed:
docs: Add tensorizer link in `engine_args.rst`, docstring to example

Added a link to tensorizer in `engine_args.rst` as well as a docstring
describing `examples/tensorize_vllm_model.py` in more detail.

Yard1 commented:
I've opened https://github.com/vllm-project/vllm/pull/3977 and I confirmed that applying it on top of this PR fixes the issue.

Yard1 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,267 @@
+                          str, bytes, os.PathLike, int]
+    verify_hash: bool = False
+    encryption_keyfile: Optional[str] = None
+    num_readers: Optional[int] = None
```
This should either default to 1 or become a mandatory argument

Yard1 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,267 @@
+    s3_access_key_id: Optional[str] = None
+    s3_secret_access_key: Optional[str] = None
+    s3_endpoint: Optional[str] = None
+    vllm_tensorized: Optional[bool] = False
```
I would suggest making this mandatory too

Yard1 commented on vllm/model_executor/model_loader.py:
```diff
@@ -77,7 +80,11 @@ def get_model(model_config: ModelConfig, device_config: DeviceConfig,
         # The weights will be initialized as empty tensors.
         with torch.device(device_config.device):
-            if hasattr(model_class, "supported_lora_modules"):
+            if model_config.load_format == "tensorizer" and _is_vllm_model(
+                    model_config):
+                model = load_with_tensorizer(model_class, model_config)
+                return model.eval()
```
we should either support LoRAs and different linear methods here, or throw exceptions if they are specified to be used with tensorizer.

I am trying out LoRA support, btw

sangstar committed:
chore: Resolve comments

Resolving comments related to making `num_readers` default set to `1`,
`vllm_tensorized` set to a required arg, and tensorizer no longer
being a mandatory dep.

Yard1 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,267 @@
+class TensorizerArgs:
+    tensorizer_uri: Union[io.BufferedIOBase, io.RawIOBase, typing.BinaryIO,
+                          str, bytes, os.PathLike, int]
+    vllm_tensorized: bool = False
```
```suggestion
    vllm_tensorized: bool
```

sangstar replied:
Whoops! My bad

Yard1 commented:
@sangstar We can easily support LoRA with the following workaround in `TensorizerAgent`:
```python
    def _resize_lora_embeddings(self):
        """Modify LoRA embedding layers to use bigger tensors
        to allow for adapter added tokens."""
        for child in self.model.modules():
            if (isinstance(child, VocabParallelEmbedding)
                    and child.weight.shape[0] <
                    child.num_embeddings_per_partition):
                new_weight = torch.empty(child.num_embeddings_per_partition,
                                         child.embedding_dim,
                                         dtype=child.weight.dtype,
                                         device=child.weight.device)
                new_weight[:child.weight.shape[0]].copy_(child.weight.data)
                new_weight[child.weight.shape[0]:].fill_(0)
                child.weight.data = new_weight
```

We also need to modify `TensorizerAgent` to take in `kwargs` to pass to the model class.

Then the model loading code can become
```python
    with _set_default_torch_dtype(model_config.dtype):
        # Create a model instance.
        # The weights will be initialized as empty tensors.
        extra_kwargs = {}
        if hasattr(model_class, "supported_lora_modules"):
            extra_kwargs["lora_config"] = lora_config
        elif lora_config:
            raise ValueError(
                f"Model {model_class.__name__} does not support LoRA, "
                "but LoRA is enabled. Support for this model may "
                "be added in the future. If this is important to you, "
                "please open an issue on github.")
        elif model_class in _VISION_MODEL_CLASSES:
            extra_kwargs["vision_language_config"] = vision_language_config

        with torch.device(device_config.device):
            if model_config.load_format == "tensorizer" and _is_vllm_model(
                    model_config):
                model = load_with_tensorizer(model_class,
                                             model_config,
                                             linear_method=linear_method,
                                             **extra_kwargs)
                return model.eval()
            model = model_class(config=model_config.hf_config,
                                linear_method=linear_method,
                                **extra_kwargs)
```

sangstar committed:
fix: Affirm mandatory `vllm_tensorized` argument change

sangstar committed:
perf: Allow preliminary support deserializing with LoRA adapters

sangstar commented:
> @sangstar We can easily support LoRA with the following workaround in `TensorizerAgent`:
> 
> ```python
>     def _resize_lora_embeddings(self):
>         """Modify LoRA embedding layers to use bigger tensors
>         to allow for adapter added tokens."""
>         for child in self.model.modules():
>             if (isinstance(child, VocabParallelEmbedding)
>                     and child.weight.shape[0] <
>                     child.num_embeddings_per_partition):
>                 new_weight = torch.empty(child.num_embeddings_per_partition,
>                                          child.embedding_dim,
>                                          dtype=child.weight.dtype,
>                                          device=child.weight.device)
>                 new_weight[:child.weight.shape[0]].copy_(child.weight.data)
>                 new_weight[child.weight.shape[0]:].fill_(0)
>                 child.weight.data = new_weight
> ```
> 
> This could theoretically lead to memory fragmentation but empirically I don't see a difference in the number of available GPU blocks.
> 
> We also need to modify `TensorizerAgent` to take in `kwargs` to pass to the model class.
> 
> Then the model loading code can become
> 
> ```python
>     with _set_default_torch_dtype(model_config.dtype):
>         # Create a model instance.
>         # The weights will be initialized as empty tensors.
>         extra_kwargs = {}
>         if hasattr(model_class, "supported_lora_modules"):
>             extra_kwargs["lora_config"] = lora_config
>         elif lora_config:
>             raise ValueError(
>                 f"Model {model_class.__name__} does not support LoRA, "
>                 "but LoRA is enabled. Support for this model may "
>                 "be added in the future. If this is important to you, "
>                 "please open an issue on github.")
>         elif model_class in _VISION_MODEL_CLASSES:
>             extra_kwargs["vision_language_config"] = vision_language_config
> 
>         with torch.device(device_config.device):
>             if model_config.load_format == "tensorizer" and _is_vllm_model(
>                     model_config):
>                 model = load_with_tensorizer(model_class,
>                                              model_config,
>                                              linear_method=linear_method,
>                                              **extra_kwargs)
>                 return model.eval()
>             model = model_class(config=model_config.hf_config,
>                                 linear_method=linear_method,
>                                 **extra_kwargs)
> ```

Wonderful! I've attempted to implement those changes [here](https://github.com/vllm-project/vllm/pull/3476/commits/f82b25a36694486c48d7e31b84f7cf44b459070c).

Yard1 commented on tests/test_tensorizer.py:
```diff
@@ -0,0 +1,116 @@
+    # Assumes SamplingParams being seeded ensures the outputs are deterministic
+    assert outputs == deserialized_outputs
+
+def test_deserialized_hf_model_has_same_outputs(hf_runner, vllm_runner, tmp_path):
```
test looks good, let's also add one for LoRA?

sangstar replied:
Absolutely! Only question is what `repo_id` I can use. The one you have is for `Llama-2-7b-hf`, which is pretty large for a test.

Yard1 commented on vllm/model_executor/weight_utils.py:
```diff
@@ -4,22 +4,26 @@
 import torch
 from huggingface_hub import HfFileSystem, snapshot_download
 from safetensors.torch import load_file, safe_open, save_file
+from tensorizer.serialization import TensorDeserializer
+from tensorizer.stream_io import open_stream
```
let's import from `vllm.model_executor.tensorizer_loader` instead

Yard1 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,294 @@
+from dataclasses import dataclass
+from typing import Optional, Type, Union
+
+from tensorizer import DecryptionParams, TensorDeserializer, stream_io
+from tensorizer.utils import convert_bytes, get_mem_usage, no_init_or_tensor
```
as tensorizer is optional, can we wrap all tensorizer imports in a try...except ImportError block and fail gracefully? we could re-raise in TensorizerAgent init

let's make sure that vllm can run without tensorizer installed!

Yard1 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,294 @@
+                                 **extra_kwargs)
+    return tensorizer.deserialize()
+
+def _is_vllm_model(model_config: ModelConfig) -> bool:
```
suggest `is_vllm_serialized_tensorizer` or something similar like that

Yard1 commented on requirements-rocm.txt:
```diff
@@ -9,6 +9,7 @@ ray == 2.9.3
 sentencepiece  # Required for LLaMA tokenizer.
 numpy
 tokenizers>=0.15.0
+tensorizer >= 2.8.1
```
remove

Yard1 commented on requirements-cpu.txt:
```diff
@@ -9,7 +9,8 @@ fastapi
 prometheus_client >= 0.18.0
-torch == 2.1.2+cpu 
+torch == 2.1.2+cpu
+tensorizer == 2.9.0a0
```
remove

sangstar committed:
fix: Fix requirements.txt passing import tensorizer only if installed

sangstar committed:
fix: Properly ensure import fail if tensorizer not used nor installed

sangstar committed:
perf: Move test location and add testing for LoRA

sangstar commented:
@Yard1 I should have a initial refactor for you using `TensorizerConfig` presently that will make things flow a lot nicer.

sangstar committed:
perf: Add some testing changes, introduce `TensorizerConfig`

sangstar committed:
chore: Add `__init__.py` for `tests/tensorizer`

sangstar committed:
tests: Fix `test_tensorizer.py` to account for new changes

sangstar committed:
tests: Remove `test_tensorizer_api_server.py`

sangstar committed:
Run yapf and ruff; fix tests

sangstar committed:
fix: Revert change to `examples/multilora_inference.py`

Yard1 commented on examples/multilora_inference.py:
```diff
@@ -103,6 +103,7 @@ def initialize_engine() -> LLMEngine:
     #   numbers will cause higher memory usage. If you know that all LoRAs will
     #   use the same rank, it is recommended to set this as low as possible.
     # max_cpu_loras: controls the size of the CPU LoRA cache.
+    from vllm.model_executor.tensorizer_loader import TensorizerArgs
```
remove?

Yard1 reviewed (commented):
Thanks, this is looking really good! Now that https://github.com/vllm-project/vllm/pull/3977 is merged can you merge master and remove the workaround?

sangstar committed:
Merge remote-tracking branch 'upstream/main' into sangstar/integrate-tensorizer

# Conflicts:
#	requirements-cpu.txt
#	vllm/executor/gpu_executor.py
#	vllm/worker/model_runner.py
#	vllm/worker/worker.py

sangstar committed:
Merge remote-tracking branch 'upstream/main' into sangstar/integrate-tensorizer

sangstar committed:
perf: Update code to reflect change in #3977

sangstar commented:
> Thanks, this is looking really good! Now that #3977 is merged can you merge master and remove the workaround?

Merged master and removed the workaround! Seems my tests broke about parallel process groups.

rkooo567 commented on docs/source/models/engine_args.rst:
```diff
@@ -45,6 +45,7 @@ Below, you can find an explanation of every engine argument for vLLM:
     * "safetensors" will load the weights in the safetensors format.
     * "npcache" will load the weights in pytorch format and store a numpy cache to speed up the loading.
     * "dummy" will initialize the weights with random values, mainly for profiling.
+    * "tensorizer" will load serialized weights using `CoreWeave's Tensorizer model deserializer. <https://github.com/coreweave/tensorizer>`_.
```
Can you add a link to the example doc here?

sangstar replied:
Do you mean `examples/tensorize_vllm_model.py`? The tensorizer repo itself is the best resource for tensorizer documentation.

rkooo567 replied:
yes!

rkooo567 replied:
I think it is almost necessary to look at that file for any user who wants to use this feature. I think you should also mention tensorizer should be downloaded with pip install

sangstar replied:
Added!

rkooo567 commented on examples/tensorize_vllm_model.py:
```diff
@@ -0,0 +1,267 @@
+   deserialize \
+   --path-to-tensors s3://my-bucket/vllm/EleutherAI/gpt-j-6B/vllm/model.tensors
+
+Which downloads the model tensors from your S3 bucket and deserializes them.
```
Can you add a short doc about how to handle S3 credentials? (does it assume env var exists, or the instance already has an access)?

rkooo567 reviewed (commented):
Some minor comments about the doc

sangstar committed:
chore: Remove accidental syntax error

sangstar committed:
docs: Elaborate on S3 credentialing

sangstar committed:
fix: Properly passing `tensorizer_config` to hf weight loader

sangstar committed:
fix: Fix, test tensorizer uri passing without tensorizer load format

sangstar committed:
docs: Note example script in docs for more information

sangstar committed:
chore: Run yapf and ruff, as well as doc edits

sangstar committed:
fix: Fix `initialize_model_parallel` import

sangstar committed:
tests: Add test for `examples/tensorize_vllm_model.py`

sangstar committed:
tests: Fix lora test

sangstar committed:
Run yapf and ruff

sangstar committed:
fix: Move `tensorize_loader` imports to pass CPU test

sangstar commented:
Added several more docs changes and tests -- including integration tests for deserializing from s3, binding LoRA adapters to a vLLM model and testing the example script serializing and deserializing successfully. yapf seems to be complaining about an extra line it wants to remove on line 17 of my example script, but I don't see the line in question.

sangstar committed:
refactor: Pass `TensorizerArgs` direct to `EngineArgs.add_cli_args`

sangstar committed:
tests: Add api_server test using tensorizer

sangstar committed:
fix: Add `tensorizer_config` to `RayGPUExecutor`

sangstar committed:
tests: Formatting and add test to ensure `tensorizer` load format

sangstar committed:
style: Run yapf on `examples/tensorize_vllm_model.py`

sangstar committed:
style: Run isort on `examples/tensorize_vllm_model.py`

sangstar committed:
style: Fix yapf and isort conflict

sangstar committed:
fix: Remove `tensorizer_args` from `ModelConfig`

sangstar commented:
Added additional tests and some fixes; checks are all passing!

rkooo567 reviewed (approved):
LGTM if @Yard1 gives a go

ywang96 commented on examples/tensorize_vllm_model.py:
```diff
@@ -0,0 +1,275 @@
+# yapf conflicts with isort for this docstring
+# yapf: disable
+"""
+tensorize_vllm_model.py is a script that can be used to serialize and 
+deserialize vLLM models. These models can be loaded using tensorizer directly 
+to the GPU extremely quickly. Tensor encryption and decryption is also 
```
Just a nit: This is slightly misleading. The serialization step is done for `AutoModelForCausalLM` model object but we verify if it can be supported by `vllm`  only in deserialization step.

sangstar replied:
The model serialized in this script is done for a model instantiated as a vLLM model.

```
def serialize():
    model = AutoModelForCausalLM.from_pretrained(model_ref)
    config = AutoConfig.from_pretrained(model_ref)
    make_model_contiguous(model)
    to_dtype = _get_and_verify_dtype(config, dtype=dtype)
    model = model.to(dtype=to_dtype)
    with tempfile.TemporaryDirectory() as tmpdir:
        DOWNLOAD_DIR = os.path.join(tmpdir, model_name)
        model.save_pretrained(DOWNLOAD_DIR)
        del model
        model_class = _get_vllm_model_architecture(config)
```

Where `_get_vllm_model_architecture` is defined as

```
def _get_vllm_model_architecture(config: PretrainedConfig) -> Type[nn.Module]:
    architectures = getattr(config, "architectures", [])
    for arch in architectures:
        model_cls = ModelRegistry.load_model_cls(arch)
        if model_cls is not None:
            return model_cls
    raise ValueError(
        f"Model architectures {architectures} are not supported for now. "
        f"Supported architectures: {ModelRegistry.get_supported_archs()}")
```

Which makes use of `ModelRegistry.load_model_cls` to instantiate a vLLM model for serialization, and loads the config into it.

ywang96 replied:
Ah I see! Sorry about not reading it carefully!

sangstar replied:
No worries! I should have all of your comments resolved with a commit coming briefly. It also uses `LLMEngine` here directly to ensure all of vLLM's model loading optimizations and architecture are used.

ywang96 commented on examples/tensorize_vllm_model.py:
```diff
@@ -0,0 +1,275 @@
+
+
+def serialize():
+    model = AutoModelForCausalLM.from_pretrained(model_ref)
```
QoL nit: How about using `huggingface_hub.snapshot_download` so the weights can be downloaded concurrently?

sangstar replied:
Will do!

ywang96 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,304 @@
+            help="The endpoint for the S3 bucket. Can also be set via the "
+            "S3_ENDPOINT_URL environment variable.",
+        )
+        group.add_argument(
+            "--vllm-tensorized",
+            action="store_true",
+            help="If enabled, indicates that the serialized model is a vLLM "
+            "model. This is used to determine the behavior of the "
+            "TensorDeserializer when loading tensors from a "
+            "serialized model.")
```
I might be missing something here, but shouldn't we always assume the tensors are serialized and verified from the example script, and thus is always vLLM model?

sangstar replied:
Not necessarily. Users of `tensorizer` have models serialized from HuggingFace like in the serialize quickstart example in the [`tensorizer` repo.](https://github.com/coreweave/tensorizer/blob/main/examples/serialize.py) We also have a `tensorized` bucket on S3 with many pre-serialized models from HuggingFace, so there are lots of tensorized HuggingFace models out there.

ywang96 commented on vllm/model_executor/tensorizer_loader.py:
```diff
@@ -0,0 +1,294 @@
+        logger.info(f"Memory usage after: {after_mem}")
+
+        self._patch_linear_weights()
+        self._resize_lora_embeddings()
```
+1

ywang96 commented on examples/tensorize_vllm_model.py:
```diff
@@ -0,0 +1,275 @@
+        help="The model reference name to serialize or deserialize. "
+        "This should be a HuggingFace ID for the model, e.g. "
+        "EleutherAI/gpt-j-6B.")
+    parser.add_argument("--dtype",
+                        type=str,
+                        default="float16",
+                        required=False,
+                        help="The dtype to cast the tensors to. ")
```
Should we infer the default dtype from the model repo instead of `float16`? For example, the default `dtype` for `mistralai/Mixtral-8x7B-Instruct-v0.1` is actually `bfloat16`.

sangstar replied:
I've removed this as a requirement, and surrender this to `dtype` from `EngineArgs`.

ywang96 reviewed (commented):
Thank you very much @sangster for your contribution to add this feature, and @Yard1 @rkooo567 for the detailed reviews!

I've tested this PR on different scenarios (from S3, from a local directory with serialized weights using the example script, etc) and left a few feedback on the PR. PTAL, thanks!

sangstar committed:
fix: Add error for device scattering and initial handling for quant

sangstar committed:
perf: Multiple changes in response to comments

Replaced the model initialization process using `LLMEngine`, allowing
vLLM to handle and therefore optimize the initial model loading
process. Added testing for quantization.

sangstar commented on vllm/engine/arg_utils.py:
```diff
@@ -465,6 +484,17 @@ def create_engine_config(self, ) -> EngineConfig:
             max_cpu_loras=self.max_cpu_loras if self.max_cpu_loras
             and self.max_cpu_loras > 0 else None) if self.enable_lora else None
 
+        tensorizer_config = TensorizerConfig(
+            tensorizer_uri=self.tensorizer_uri,
+            vllm_tensorized=self.vllm_tensorized,
+            verify_hash=self.verify_hash,
+            num_readers=self.num_readers,
+            encryption_keyfile=self.encryption_keyfile,
+            s3_access_key_id=self.s3_access_key_id,
+            s3_secret_access_key=self.s3_secret_access_key,
+            s3_endpoint=self.s3_endpoint,
+        )
+
```
I'll have it throw an error. Tensorizer can theoretically handle multiple devices through the use of multiple deserializers, but there isn't any official support for it at the moment so I think, if at all, it deserves to have its own PR in the future.

ywang96 replied:
Yep that works too!

sangstar committed:
perf: Final changes to resolve comments

sangstar commented:
@ywang96 @Yard1 @rkooo567 

Thank you all very much for your reviews! I've implemented the changes from @ywang96 's comments.

sangstar committed:
fix: Skip tests if cURL not installed, add example script for testing

sangstar committed:
Run yapf and ruff

sangstar committed:
tests: Install cURL for tensorizer tests for testing suite

sangstar committed:
tests: Install libsodium23 for CI tensorizer tests

sangstar committed:
fix: Fix testing import path

sangstar committed:
Run yapf and ruff

sangstar commented:
Some minor fixes to ensure the testing suite can run the tensorizer tests. All passing! Thanks very much again for the reviews @rkooo567 @Yard1 @ywang96 let me know if anything else needed! :)

ywang96 reviewed (approved):
LGTM! Thank you again @sangstar for all the work and test coverage on this PR to add this feature!

bbrowning commented:
I was able to successfully test this in a vLLM 0.4.1 container running on OpenShift, both with models serialized with the Tensorizer library directly and for vLLM-serialized models. Once I cranked up the Pod's CPU and increased the `num_readers` parameter, I got about an 8x speedup in my case when loading the same model via vLLM-serialized tensorize files compared to not using tensorizer at all and just downloading the safetensors from S3 to local disk then loading with vLLM. This took my overall cold start time of this Pod from a bit over 4 minutes to 30 seconds. There may be even more performance available in my setup with additional tweaking, but this is already a great win.

`INFO 05-03 11:11:38 tensorizer.py:337] Deserialized 14.5 GB in 15.21s, 953.1 MB/s`

That's an awesome improvement, and thank you!

sangstar commented:
> I was able to successfully test this in a vLLM 0.4.1 container running on OpenShift, both with models serialized with the Tensorizer library directly and for vLLM-serialized models. Once I cranked up the Pod's CPU and increased the `num_readers` parameter, I got about an 8x speedup in my case when loading the same model via vLLM-serialized tensorize files compared to not using tensorizer at all and just downloading the safetensors from S3 to local disk then loading with vLLM. This took my overall cold start time of this Pod from a bit over 4 minutes to 30 seconds. There may be even more performance available in my setup with additional tweaking, but this is already a great win.
> 
> `INFO 05-03 11:11:38 tensorizer.py:337] Deserialized 14.5 GB in 15.21s, 953.1 MB/s`
> 
> That's an awesome improvement, and thank you!

I'm thrilled to hear that! I currently actually have a new PR up #4208 that uses the full `2.9.0` release, has better usage documentation, and automated inferring a vLLM-serialized model.

nskumz commented:
`root@benchmark-model-8b-79687c7db6-ndpqq:/huggingface#` python3 -m vllm.entrypoints.openai.api_server --host 0.0.0.0 --model /huggingface/llama-3-1-8b-instruct --load-format tensorizer --tensorizer-uri /huggingface/llama-3-1-8b-instruct/llama-3-1-8b-instruct.tensors
usage: api_server.py [-h] [--host HOST] [--port PORT] [--uvicorn-log-level {debug,info,warning,error,critical,trace}] [--allow-credentials] [--allowed-origins ALLOWED_ORIGINS] [--allowed-methods ALLOWED_METHODS]
                     [--allowed-headers ALLOWED_HEADERS] [--api-key API_KEY] [--lora-modules LORA_MODULES [LORA_MODULES ...]] [--prompt-adapters PROMPT_ADAPTERS [PROMPT_ADAPTERS ...]]
                     [--chat-template CHAT_TEMPLATE] [--chat-template-content-format {auto,string,openai}] [--response-role RESPONSE_ROLE] [--ssl-keyfile SSL_KEYFILE] [--ssl-certfile SSL_CERTFILE]
                     [--ssl-ca-certs SSL_CA_CERTS] [--ssl-cert-reqs SSL_CERT_REQS] [--root-path ROOT_PATH] [--middleware MIDDLEWARE] [--return-tokens-as-token-ids] [--disable-frontend-multiprocessing]
                     [--enable-request-id-headers] [--enable-auto-tool-choice]
                     [--tool-call-parser {granite-20b-fc,granite,hermes,internlm,jamba,llama3_json,mistral,pythonic} or name registered in --tool-parser-plugin] [--tool-parser-plugin TOOL_PARSER_PLUGIN]
                     [--model MODEL] [--task {auto,generate,embedding,embed,classify,score,reward}] [--tokenizer TOKENIZER] [--skip-tokenizer-init] [--revision REVISION] [--code-revision CODE_REVISION]
                     [--tokenizer-revision TOKENIZER_REVISION] [--tokenizer-mode {auto,slow,mistral}] [--trust-remote-code] [--allowed-local-media-path ALLOWED_LOCAL_MEDIA_PATH] [--download-dir DOWNLOAD_DIR]
                     [--load-format {auto,pt,safetensors,npcache,dummy,tensorizer,sharded_state,gguf,bitsandbytes,mistral,runai_streamer}] [--config-format {auto,hf,mistral}]
                     [--dtype {auto,half,float16,bfloat16,float,float32}] [--kv-cache-dtype {auto,fp8,fp8_e5m2,fp8_e4m3}] [--quantization-param-path QUANTIZATION_PARAM_PATH] [--max-model-len MAX_MODEL_LEN]
                     [--guided-decoding-backend {outlines,lm-format-enforcer,xgrammar}] [--logits-processor-pattern LOGITS_PROCESSOR_PATTERN] [--distributed-executor-backend {ray,mp}] [--worker-use-ray]
                     [--pipeline-parallel-size PIPELINE_PARALLEL_SIZE] [--tensor-parallel-size TENSOR_PARALLEL_SIZE] [--max-parallel-loading-workers MAX_PARALLEL_LOADING_WORKERS] [--ray-workers-use-nsight]
                     [--block-size {8,16,32,64,128}] [--enable-prefix-caching | --no-enable-prefix-caching] [--disable-sliding-window] [--use-v2-block-manager] [--num-lookahead-slots NUM_LOOKAHEAD_SLOTS]
                     [--seed SEED] [--swap-space SWAP_SPACE] [--cpu-offload-gb CPU_OFFLOAD_GB] [--gpu-memory-utilization GPU_MEMORY_UTILIZATION] [--num-gpu-blocks-override NUM_GPU_BLOCKS_OVERRIDE]
                     [--max-num-batched-tokens MAX_NUM_BATCHED_TOKENS] [--max-num-seqs MAX_NUM_SEQS] [--max-logprobs MAX_LOGPROBS] [--disable-log-stats]
                     [--quantization {aqlm,awq,deepspeedfp,tpu_int8,fp8,fbgemm_fp8,modelopt,marlin,gguf,gptq_marlin_24,gptq_marlin,awq_marlin,gptq,compressed-tensors,bitsandbytes,qqq,hqq,experts_int8,neuron_quant,ipex,None}]
                     [--rope-scaling ROPE_SCALING] [--rope-theta ROPE_THETA] [--hf-overrides HF_OVERRIDES] [--enforce-eager] [--max-seq-len-to-capture MAX_SEQ_LEN_TO_CAPTURE] [--disable-custom-all-reduce]
                     [--tokenizer-pool-size TOKENIZER_POOL_SIZE] [--tokenizer-pool-type TOKENIZER_POOL_TYPE] [--tokenizer-pool-extra-config TOKENIZER_POOL_EXTRA_CONFIG] [--limit-mm-per-prompt LIMIT_MM_PER_PROMPT]
                     [--mm-processor-kwargs MM_PROCESSOR_KWARGS] [--disable-mm-preprocessor-cache] [--enable-lora] [--enable-lora-bias] [--max-loras MAX_LORAS] [--max-lora-rank MAX_LORA_RANK]
                     [--lora-extra-vocab-size LORA_EXTRA_VOCAB_SIZE] [--lora-dtype {auto,float16,bfloat16}] [--long-lora-scaling-factors LONG_LORA_SCALING_FACTORS] [--max-cpu-loras MAX_CPU_LORAS]
                     [--fully-sharded-loras] [--enable-prompt-adapter] [--max-prompt-adapters MAX_PROMPT_ADAPTERS] [--max-prompt-adapter-token MAX_PROMPT_ADAPTER_TOKEN]
                     [--device {auto,cuda,neuron,cpu,openvino,tpu,xpu,hpu}] [--num-scheduler-steps NUM_SCHEDULER_STEPS] [--multi-step-stream-outputs [MULTI_STEP_STREAM_OUTPUTS]]
                     [--scheduler-delay-factor SCHEDULER_DELAY_FACTOR] [--enable-chunked-prefill [ENABLE_CHUNKED_PREFILL]] [--speculative-model SPECULATIVE_MODEL]
                     [--speculative-model-quantization {aqlm,awq,deepspeedfp,tpu_int8,fp8,fbgemm_fp8,modelopt,marlin,gguf,gptq_marlin_24,gptq_marlin,awq_marlin,gptq,compressed-tensors,bitsandbytes,qqq,hqq,experts_int8,neuron_quant,ipex,None}]
                     [--num-speculative-tokens NUM_SPECULATIVE_TOKENS] [--speculative-disable-mqa-scorer] [--speculative-draft-tensor-parallel-size SPECULATIVE_DRAFT_TENSOR_PARALLEL_SIZE]
                     [--speculative-max-model-len SPECULATIVE_MAX_MODEL_LEN] [--speculative-disable-by-batch-size SPECULATIVE_DISABLE_BY_BATCH_SIZE] [--ngram-prompt-lookup-max NGRAM_PROMPT_LOOKUP_MAX]
                     [--ngram-prompt-lookup-min NGRAM_PROMPT_LOOKUP_MIN] [--spec-decoding-acceptance-method {rejection_sampler,typical_acceptance_sampler}]
                     [--typical-acceptance-sampler-posterior-threshold TYPICAL_ACCEPTANCE_SAMPLER_POSTERIOR_THRESHOLD] [--typical-acceptance-sampler-posterior-alpha TYPICAL_ACCEPTANCE_SAMPLER_POSTERIOR_ALPHA]
                     [--disable-logprobs-during-spec-decoding [DISABLE_LOGPROBS_DURING_SPEC_DECODING]] [--model-loader-extra-config MODEL_LOADER_EXTRA_CONFIG] [--ignore-patterns IGNORE_PATTERNS]
                     [--preemption-mode PREEMPTION_MODE] [--served-model-name SERVED_MODEL_NAME [SERVED_MODEL_NAME ...]] [--qlora-adapter-name-or-path QLORA_ADAPTER_NAME_OR_PATH]
                     [--otlp-traces-endpoint OTLP_TRACES_ENDPOINT] [--collect-detailed-traces COLLECT_DETAILED_TRACES] [--disable-async-output-proc] [--scheduling-policy {fcfs,priority}]
                     [--override-neuron-config OVERRIDE_NEURON_CONFIG] [--override-pooler-config OVERRIDE_POOLER_CONFIG] [--compilation-config COMPILATION_CONFIG] [--kv-transfer-config KV_TRANSFER_CONFIG]
                     [--worker-cls WORKER_CLS] [--generation-config GENERATION_CONFIG] [--disable-log-requests] [--max-log-len MAX_LOG_LEN] [--disable-fastapi-docs] [--enable-prompt-tokens-details]
api_server.py: error: unrecognized arguments: --tensorizer-uri /huggingface/llama-3-1-8b-instruct/llama-3-1-8b-instruct.tensors


so, i tried recently on vLLM v0.6.6 vesrion, with help of tensorizer repo i serialized the model and placed in the folder( doing in local not s3) 
so the thing is i searched for required argument [[[--tensorizer-uri]]] with vllm serve -h command ( i havent found the argument)

@sangstar could you please help me out of this
